### PR TITLE
Improve Assembly (De)Serialization

### PR DESCRIPTION
**Pull request by itzmeanjan**
This PR addresses the issue described on https://github.com/0xPolygonMiden/miden-vm/issues/731.

- [x] Make Miden assembly serialization infallible
- [x] Use Serializable/Deserializable traits defined in `winter-utils` ( more @ https://github.com/facebook/winterfell/blob/9b33ffe63ce48385349381f25341111376de19bd/utils/core/src/serde/mod.rs ) for reducing code duplication in `miden-assembly` ( more @ https://github.com/0xPolygonMiden/miden-vm/blob/4195475d75ab2d586bdb01d1ff3ea2cd626eaf7b/assembly/src/serde/mod.rs )

Run tests by issuing

```bash
make test
```

### CODE REVIEW

File: assembly/src/serde/mod.rs [hunk 1]
```diff
@@ -33,22 +33,13 @@ impl ByteWriter {
...
         self.write_u64(val.as_int());
     }
 
-    pub fn write_str(&mut self, val: &str) -> Result<(), SerializationError> {
-        self.write_len(val.len())?;
+    pub fn write_str(&mut self, val: &str) {
+        self.write_u64(val.len() as u64);
         self.0.extend(val.as_bytes());
-        Ok(())
     }
```
Comment (bobbinth) [inline assembly/src/serde/mod.rs:43]: i would remove this method as winterfell `Serializable` trait doesn't have it - and so, we'd need to remove it anyway.

The way to handle this is to handle string serialization at the call sites of this method.

File: assembly/src/serde/mod.rs [hunk 2]
```diff
@@ -193,7 +177,7 @@ impl<'a> ByteReader<'a> {
...
 
     pub fn read_str(&mut self) -> Result<&str, SerializationError> {
-        let len = self.read_u16()? as usize;
+        let len = self.read_u64()? as usize;
         self.check_eor(len)?;
         let string = &self.bytes[self.pos..self.pos + len];
         self.pos += len;
```
Comment (bobbinth) [inline assembly/src/serde/mod.rs:183]: Similar to the previous comment. I would remove this method.

File: assembly/src/serde/mod.rs [hunk 3]
```diff
@@ -69,56 +60,53 @@ impl ByteWriter {
...
     }
 }
 
 impl Serializable for String {
-    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
-        target.write_str(self)
+    fn write_into(&self, target: &mut ByteWriter) {
+        target.write_str(self);
     }
 }
```
Comment (bobbinth) [inline assembly/src/serde/mod.rs:86]: I would remove these implementations. And also would remove implementations for serializing `Vec<T>` below.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left a few small comment inline.

Main thing: we should remove generic serializers for strings and vectors and handle these serializations individually at their call sites.

**Comment by itzmeanjan** [comment]
> Main thing: we should remove generic serializers for strings and vectors and handle these serializations individually at their call sites.

I make those changes in https://github.com/0xPolygonMiden/miden-vm/pull/825/commits/66523999f6631d2679548ea19fb63f2d82dedd4e. Now we're only dependent on `winter-utils` crate's (De)Serializable interfaces.

File: assembly/src/library.rs [hunk 4]
```diff
@@ -230,16 +230,15 @@ mod use_std {
...
-        modules.try_for_each(|module| {
-            ModulePath::strip_namespace(&module.path).write_into(target)?;
-            module.ast.write_into(target)?;
-            Ok(())
+    fn write_into(&self, target: &mut ByteWriter) {
+        self.namespace.write_into(target);
+        self.version.write_into(target);
+
+        let modules = self.modules();
+        target.write_u64(modules.len() as u64);
```
[the remark this answers is not in the document]
Comment (itzmeanjan) [reply assembly/src/library.rs:238]: I use `u16` for encoding length of string for certain cases, though generally I stick to using `u64`. I hope I've not missed any.

File: assembly/src/parsers/mod.rs [hunk 5]
```diff
@@ -142,26 +147,47 @@ pub struct ProcedureAst {
...
+            None => {
+                target.write_bool(false);
+            }
+        }
         target.write_bool(self.is_export);
         target.write_u16(self.num_locals);
-        self.body.write_into(target)?;
-        Ok(())
+        target.write_u64(self.body.len() as u64);
+        self.body.write_into(target);
```
Comment (bobbinth) [inline assembly/src/parsers/mod.rs:165]: I think `u64` both for docs and for body length is too much - let's use `u16` and add TODOs to enforce these restrictions via the constructor.

File: assembly/src/parsers/mod.rs [hunk 6]
```diff
@@ -43,31 +43,27 @@ pub struct ProgramAst {
...
+        let mut target = Vec::<u8>::default();
 
-        // local procedures
-        byte_writer.write_u16(self.local_procs.len() as u16);
+        target.write_u16(self.local_procs.len() as u16);
+        self.local_procs.write_into(&mut target);
 
-        self.local_procs.iter().try_for_each(|proc| proc.write_into(&mut byte_writer))?;
+        target.write_u64(self.body.len() as u64);
+        self.body.write_into(&mut target);
```
Comment (bobbinth) [inline assembly/src/parsers/mod.rs:53]: I think `u64` both for body length is too much - let's use `u16` and add a TODO to enforce this restriction via the constructor.

File: assembly/src/parsers/mod.rs [hunk 7]
```diff
@@ -85,44 +81,53 @@ pub struct ModuleAst {
...
+            Some(t) => {
+                target.write_bool(true);
+                target.write_u64(t.len() as u64);
+                target.write_bytes(t.as_bytes());
+            }
+            None => {
+                target.write_bool(false);
+            }
+        }
+        target.write_u64(self.local_procs.len() as u64);
```
Comment (bobbinth) [inline assembly/src/parsers/mod.rs:109]: I think `u64` both for docs and for the number of procedures is too much - let's use `u16` and add TODOs to enforce these restrictions via the constructor.

File: assembly/src/parsers/serde/deserialization.rs [hunk 8]
```diff
@@ -1,29 +1,43 @@
...
-                Deserializable::read_from(bytes)?,
-                Deserializable::read_from(bytes)?,
-            ))
+            source.read_u8()?;
+
+            let if_block_len = source.read_u64()? as usize;
+            let if_block = Deserializable::read_batch_from(source, if_block_len)?;
+
+            let else_block_len = source.read_u64()? as usize;
+            let else_block = Deserializable::read_batch_from(source, else_block_len)?;
```
Comment (bobbinth) [inline assembly/src/parsers/serde/deserialization.rs:20]: I think `u64` for number of nodes here and for other statements is too much - let's use `u16` and add TODOs to enforce these restrictions via the constructor.

File: assembly/src/parsers/serde/deserialization.rs [hunk 9]
```diff
@@ -248,74 +262,74 @@ impl Deserializable for Instruction {
...
             OpCode::PushWord => Ok(Instruction::PushWord([
-                bytes.read_felt()?,
-                bytes.read_felt()?,
-                bytes.read_felt()?,
-                bytes.read_felt()?,
+                Felt::read_from(source)?,
+                Felt::read_from(source)?,
+                Felt::read_from(source)?,
+                Felt::read_from(source)?,
             ])),
```
Comment (bobbinth) [inline assembly/src/parsers/serde/deserialization.rs:274]: Would `source.read_array()?` work here instead?

File: assembly/src/library.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply assembly/src/library.rs:238]: I'd use a `u16` here still and add a TODO to ensure this restriction via the constructor.

File: assembly/src/lib.rs [hunk 10]
```diff
@@ -163,18 +164,22 @@ impl AsRef<str> for AbsolutePath {
 }
 
 impl Serializable for AbsolutePath {
-    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
-        target.write_str(&self.path)
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_u64(self.len() as u64);
+        target.write_bytes(self.as_bytes());
     }
```
Comment (bobbinth) [inline assembly/src/lib.rs:170]: I think `u64` is too much - let's use `u16` and add TODOs to enforce this restriction via the constructor.

File: assembly/src/lib.rs [hunk 11]
```diff
@@ -234,15 +239,19 @@ impl AsRef<str> for LibraryNamespace {
 }
 
 impl Serializable for LibraryNamespace {
-    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
-        target.write_str(&self.name)
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_u64(self.name.len() as u64);
+        target.write_bytes(self.name.as_bytes());
     }
```
Comment (bobbinth) [inline assembly/src/lib.rs:245]: I think `u64` is too much - let's use `u16` and add TODOs to enforce this restriction via the constructor.

File: assembly/src/lib.rs [hunk 12]
```diff
@@ -349,15 +358,19 @@ impl AsRef<str> for ModulePath {
 }
 
 impl Serializable for ModulePath {
-    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
-        target.write_str(&self.path)
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_u64(self.path.len() as u64);
+        target.write_bytes(self.path.as_bytes());
     }
```
Comment (bobbinth) [inline assembly/src/lib.rs:364]: I think `u64` is too much - let's use `u16` and add TODOs to enforce this restriction via the constructor.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left a bunch of small comments - primarily about switching length encodings from `u64` to `u16`. We should also create an issue to address the TODOs and make sure we enforce proper bounds on these at construction time.

File: assembly/src/parsers/serde/deserialization.rs (hunk 9 above)
Comment (itzmeanjan) [reply assembly/src/parsers/serde/deserialization.rs:274]: I see `read_array` reads N ( compile-time defined ) bytes from source byte stream. More @ https://github.com/facebook/winterfell/blob/9b33ffe63ce48385349381f25341111376de19bd/utils/core/src/serde/byte_reader.rs#L155-L161

File: assembly/Cargo.toml [hunk 13]
```diff
@@ -22,3 +22,4 @@ std = ["vm-core/std"]
 [dependencies]
 num_enum = "0.5.10"
 vm-core = { package = "miden-core", path = "../core", version = "0.6", default-features = false }
+winter-utils = { package = "winter-utils", version = "0.6", default-features = false }
```
[the remark this answers is not in the document]
Comment (itzmeanjan) [reply assembly/Cargo.toml]: > I think this is primarily used for serialization traits, right? If so, we can get these from [miden-core](https://github.com/0xPolygonMiden/miden-vm/blob/next/core/src/utils/mod.rs#L14) (they are re-exported there from `winter-utils`).

I address this issue in 0d14aa953fa785e1d0f253cf2fdf0bb8dfd5ae33

File: assembly/src/procedures/mod.rs [hunk 14]
```diff
@@ -160,27 +160,25 @@ impl AsRef<str> for ProcedureName {
...
-        let name = from_utf8(name_bytes).map_err(|_| SerializationError::InvalidUtf8)?;
-        let name = ProcedureName::try_from(name.to_string())?;
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let nlen = source.read_u8()? as usize;
+        let name = source.read_vec(nlen)?;
+        let name =
+            from_utf8(&name).map_err(|e| DeserializationError::InvalidValue(e.to_string()))?;
+        let name = ProcedureName::try_from(name.to_string())
+            .map_err(|e| DeserializationError::InvalidValue(e.to_string()))?;
         Ok(name)
```
[the remark this answers is not in the document]
Comment (itzmeanjan) [reply assembly/src/procedures/mod.rs]: Addressed in c038d78fc458a9c5fc5b4fbf67df2513556d8b66

**Comment by itzmeanjan** [comment]
> create an issue to address the TODOs

Created https://github.com/0xPolygonMiden/miden-vm/issues/836

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**